common/src/main/java/com/microsoft/identity/common/internal/eststelemetry/EstsTelemetry.java

```diff
@@ -194,122 +215,143 @@ private IRequestTelemetryCache createLastRequestTelemetryCache(@NonNull final Co
+        } else if (command instanceof TokenCommand) {
+            final ILocalAuthenticationResult localAuthenticationResult =
+                    (ILocalAuthenticationResult) commandResult.getResult();
+
+            if (localAuthenticationResult.isServicedFromCache()) {
+                // we returned a token from cache, let's increment the silent success count
+                lastRequestTelemetry.incrementSilentSuccessCount();
+            }
+        } // else leave everything as is
+
+        // we're done processing telemetry for this command, let's remove it from the map
         mTelemetryMap.remove(correlationId);
 
-        if (mLastRequestTelemetryCache == null) {
+        if (mLastRequestTelemetryCache != null) {
+            // save the (updated) telemetry object back to telemetry cache
+            mLastRequestTelemetryCache.saveRequestTelemetryToCache(lastRequestTelemetry);
+        } else {
             Logger.warn(
                     TAG + methodName,
                     "Last Request Telemetry Cache object was null. " +
                             "Unable to save request telemetry to cache."
             );
-        } else if (eligibleToCache(lastTelemetry)) {
-            // remove old last request telemetry data from cache
-            mLastRequestTelemetryCache.clearAll();
-            // save new last request telemetry data to cache
-            mLastRequestTelemetryCache.saveRequestTelemetryToCache(lastTelemetry);
         }
     }
 
-    // if we don't have api id then we won't save telemetry to cache
-    // this can happen for commands like the GetDeviceModeCommand
-    // that are generated via a method for which we don't want telemetry
-    private boolean eligibleToCache(RequestTelemetry lastTelemetry) {
-        return !TextUtils.isEmpty(lastTelemetry.getSchemaVersion()) &&
-                !TextUtils.isEmpty(lastTelemetry.getCommonTelemetry().get(Schema.Key.API_ID));
+    @Nullable
+    private String getErrorFromCommandResult(final CommandResult commandResult) {
+        if (commandResult.getStatus() == CommandResult.ResultStatus.ERROR) {
+            final BaseException baseException = (BaseException) commandResult.getResult();
+            return baseException.getErrorCode();
+        } else if (commandResult.getStatus() == CommandResult.ResultStatus.CANCEL) {
+            return "user_cancel";
+        } else {
+            return null;
+        }
+    }
+
+    private boolean isTelemetryLoggedByServer(@NonNull final BaseCommand command, @NonNull final CommandResult commandResult) {
+        // This was a local operation - we didn't reach token endpoint and hence telemetry wasn't sent
+        if (!(command instanceof TokenCommand)) {
+            return false;
+        }
+
+        if (commandResult.getStatus() == CommandResult.ResultStatus.ERROR) {
+            BaseException baseException = (BaseException) commandResult.getResult();
+            if (!(baseException instanceof ServiceException)) {
```

Comment:
How do we know if this is a service exception from token endpoint (not auth endpoint)?